### PR TITLE
In checkout, shipping method is not displayed if cart is only virtual

### LINKED ISSUE #18215: Shipping method choice is displayed when a cart contains a pack of virtuals products
#### Describe the bug

If we add only virtual product to the cart and checkout the "shipping method choice" is not displayed that's the good behaviour . But if the virtuals products are in a pack the "shipping method choice" is displayed

#### Expected behavior

As the pack contains only virtual products therefore the cart contains only virtual products and the choice of delivery method should not be displayed

#### Steps to Reproduce

1. Go to BO and create a pack that contains only virtuals products
2. Go to FO and add the previously created pack in the cart and proceed to checkout

**Screenshots**

virtuals product pack cart issue

#### Additional information

* PrestaShop version: 1.7.7
* PHP version: 7.2

### LINKED ISSUE #24850: Can't download Virtual Products in a Pack of Virtual Products
#### Describe the bug

When I create a Pack of Virtual Products, I can't download any of them. They are considered "normal" products.

#### Expected behavior

As a customer I should be able to Download my Virtual products

#### Steps to Reproduce

1. Create a Pack of product
2. Add 2 (or more) Virtual Products
3. Add to cart
4. In the customer account and email, customer never receive his Virtual Products

**Screenshots**

**_PS develop_**
Capture d’écran 2021-06-07 à 16 51 31

#### Additional information

* PrestaShop version: develop
* PHP version: 7.2

### PR DESCRIPTION
**Pull request by Progi1984**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | In checkout, shipping method is not displayed if cart is only virtual
| Type?             | bug fix
| Category?         | FO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #18215
| How to test?      | Cf #18215

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/25564)

### CODE REVIEW

File: classes/Cart.php [hunk 1]
```diff
@@ -4223,14 +4223,25 @@ public function hasProducts()
...
     {
-        return (bool) Db::getInstance(_PS_USE_SQL_SLAVE_)->getValue(
-            'SELECT 1 FROM ' . _DB_PREFIX_ . 'cart_product cp ' .
-            'INNER JOIN ' . _DB_PREFIX_ . 'product p
-                ON (p.is_virtual = 0 AND p.id_product = cp.id_product) ' .
-            'INNER JOIN ' . _DB_PREFIX_ . 'product_shop ps
-                ON (ps.id_shop = cp.id_shop AND ps.id_product = p.id_product) ' .
-            'WHERE cp.id_cart=' . (int) $this->id
-        );
+        // Check for not virtual products which are not packs
```
Comment (matthieu-rolland) [inline classes/Cart.php]:
```suggestion
        // Check for non-virtual products which are not packs
```

File: classes/Cart.php [hunk 2]
```diff
@@ -4223,14 +4223,25 @@ public function hasProducts()
...
+        $sql = 'SELECT 1 FROM %scart_product cp
+            INNER JOIN %sproduct p ON (p.id_product = cp.id_product AND cache_is_pack = 0 and p.is_virtual = 0)
+            INNER JOIN %sproduct_shop ps ON (ps.id_shop = cp.id_shop AND ps.id_product = p.id_product)
+            WHERE cp.id_cart=%d';
+        $sql = sprintf($sql, _DB_PREFIX_, _DB_PREFIX_, _DB_PREFIX_, (int) $this->id);
+        if ((bool) Db::getInstance(_PS_USE_SQL_SLAVE_)->getValue($sql)) {
+            return true;
+        }
+
+        // Check for not virtual product which are in packs
```
Comment (matthieu-rolland) [inline classes/Cart.php]:
```suggestion
        // Check for non-virtual products which are in packs
```

**Review by matthieu-rolland — DISMISSED** [review]
minor comment feedback, LGTM otherwise

**Review by sowbiba — DISMISSED** [review]

**Review by matks — DISMISSED** [review]

File: classes/Cart.php [hunk 3]
```diff
@@ -4223,14 +4223,25 @@ public function hasProducts()
...
-            'INNER JOIN ' . _DB_PREFIX_ . 'product_shop ps
-                ON (ps.id_shop = cp.id_shop AND ps.id_product = p.id_product) ' .
-            'WHERE cp.id_cart=' . (int) $this->id
-        );
+        // Check for non-virtual products which are not packs
+        $sql = 'SELECT 1 FROM %scart_product cp
+            INNER JOIN %sproduct p ON (p.id_product = cp.id_product AND cache_is_pack = 0 and p.is_virtual = 0)
+            INNER JOIN %sproduct_shop ps ON (ps.id_shop = cp.id_shop AND ps.id_product = p.id_product)
+            WHERE cp.id_cart=%d';
+        $sql = sprintf($sql, _DB_PREFIX_, _DB_PREFIX_, _DB_PREFIX_, (int) $this->id);
```
Comment (PierreRambaud) [inline classes/Cart.php]:
```suggestion
        $sql = sprintf($sql, _DB_PREFIX_, _DB_PREFIX_, _DB_PREFIX_, $this->id);
```

Duplicate cast already done by `%d`

File: classes/Cart.php [hunk 4]
```diff
@@ -4223,14 +4223,25 @@ public function hasProducts()
...
+            return true;
+        }
+
+        // Check for non-virtual products which are in packs
+        $sql = 'SELECT 1 FROM %scart_product cp
+            INNER JOIN %spack pa ON (pa.id_product_pack = cp.id_product)
+            INNER JOIN %sproduct p ON (p.id_product = pa.id_product_item AND p.is_virtual = 0)
+            INNER JOIN %sproduct_shop ps ON (ps.id_shop = cp.id_shop AND ps.id_product = p.id_product)
+            WHERE cp.id_cart=%d';
+        $sql = sprintf($sql, _DB_PREFIX_, _DB_PREFIX_, _DB_PREFIX_, _DB_PREFIX_, (int) $this->id);
```
Comment (PierreRambaud) [inline classes/Cart.php]:
```suggestion
        $sql = sprintf($sql, _DB_PREFIX_, _DB_PREFIX_, _DB_PREFIX_, _DB_PREFIX_, $this->id);
```
Duplicate cast already done by `%d`

Comment (Progi1984) [reply classes/Cart.php]: Fixed. Thanks

File: classes/Cart.php (hunk 3 above)
Comment (Progi1984) [reply classes/Cart.php]: Fixed. Thanks

**Review by sowbiba — APPROVED** [review]

**Review by PierreRambaud — APPROVED** [review]

**Comment by florine2623** [comment]
Hello @Progi1984 ,

It's QA approved, thanks for your PR !

**Comment by Progi1984** [comment]
Thanks @florine2623

**Pull request merged**